File: src/bcd_web_vue/js/models/borrower.js
/**
 * Borrower data models with TypeScript JSDoc type definitions.
 * Fully aligned with Pydantic and SQLAlchemy models.
 */

/**
 * Current loan/circulation transaction details
 * @typedef {Object} CurrentLoan
 * @property {string} item_id - Item identifier (barcode)
 * @property {number} bibliographic_record_id - Bibliographic record ID
 * @property {string} title - Book title
 * @property {string} author - Book author
 * @property {string} barcode - Item barcode
 * @property {string} due_date - Due date (ISO 8601 format)
 * @property {boolean} is_overdue - Whether the item is overdue
 * @property {number} days_overdue - Number of days overdue (0 if not overdue)
 * @property {number} renewals_count - Number of times renewed
 * @property {number} max_renewals - Maximum renewals allowed
 */

/**
 * Basic borrower information
 * @typedef {Object} Borrower
 * @property {number} id - Database auto-increment ID
 * @property {string} borrower_id - Unique alphanumeric borrower ID
 * @property {string} first_name - First name
 * @property {string} last_name - Last name
 * @property {string} full_name - Full name (LAST First)
 * @property {string} role - Role: 'student', 'teacher', 'staff'
 * @property {number|null} class_id - Class database ID
 * @property {string|null} grade_level - Grade level (e.g., 'CP', 'CE1')
 * @property {string} barcode - Borrower barcode (e.g., '%101')
 * @property {boolean} active - Whether borrower is active (is_active in legacy, active in backend)
 * @property {string|null} blocked_reason - Reason why the borrower is blocked, if any
 * @property {string|null} email - Email address
 * @property {string|null} phone - Phone number
 * @property {string|null} notes - Additional notes
 */

/**
 * Detailed borrower information with current loans and stats
 * @typedef {Object} BorrowerDetailed
 * @property {number} id - Database ID
 * @property {string} borrower_id - Borrower identifier
 * @property {string} first_name - First name
 * @property {string} last_name - Last name
 * @property {string} full_name - Full name
 * @property {string} role - Role
 * @property {number|null} class_id - Class database ID
 * @property {string|null} grade_level - Grade level
 * @property {string} barcode - Borrower barcode
 * @property {boolean} active - Whether borrower is active
 * @property {string|null} blocked_reason - Reason for blocking (if blocked)
 * @property {string|null} email - Email address
 * @property {string|null} phone - Phone number
 * @property {string|null} notes - Additional notes
 * @property {number} current_loans_count - Current number of items on loan
 * @property {number} total_checkouts - Total number of checkouts (all time)
 * @property {number} overdue_count - Number of overdue items
 * @property {number} loan_limit - Maximum loans allowed
 * @property {number} loan_limit_warning - Soft warning threshold for loans
 * @property {string|null} class_name - Class name (if student)
 * @property {string|null} homeroom_teacher - Homeroom teacher for the class
 * @property {CurrentLoan[]} [loans] - Array of current active loans (if returned by detail endpoint)
 */

// Export empty object to make this a module
export {};

File: src/bcd_api/schemas/borrower.py
"""Pydantic schemas for Borrower model."""

from typing import Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator

from src.bcd_api.schemas.common import TimestampMixin
from src.shared.constants import BorrowerRole


class BorrowerBase(BaseModel):
    """Base schema for Borrower."""

    borrower_id: str = Field(..., min_length=1, max_length=20, description="Unique borrower ID")
    first_name: str = Field(..., min_length=1, max_length=100, description="First name")
    last_name: str = Field(..., min_length=1, max_length=100, description="Last name")
    role: BorrowerRole = Field(..., description="Borrower role (student/teacher/staff)")
    class_id: Optional[int] = Field(None, description="Class ID (for students)")
    grade_level: Optional[str] = Field(None, max_length=20, description="Grade level")
    email: Optional[str] = Field(None, max_length=100, description="Email address")
    phone: Optional[str] = Field(None, max_length=20, description="Phone number")
    notes: Optional[str] = Field(None, description="Additional notes")


class BorrowerCreate(BorrowerBase):
    """Schema for creating a new borrower."""

    @field_validator('role', mode='before')
    @classmethod
    def validate_role(cls, v):
        """Validate and convert role to enum."""
        if isinstance(v, str):
            return BorrowerRole(v)
        return v

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "borrower_id": "101",
                "first_name": "Amira",
                "last_name": "BENALI",
                "role": "student",
                "class_id": 1,
                "grade_level": "CP",
            }
        }
    )


class BorrowerUpdate(BaseModel):
    """Schema for updating a borrower."""

    borrower_id: Optional[str] = Field(None, min_length=1, max_length=20, description="New borrower ID")
    first_name: Optional[str] = Field(None, min_length=1, max_length=100)
    last_name: Optional[str] = Field(None, min_length=1, max_length=100)
    role: Optional[BorrowerRole] = None
    class_id: Optional[int] = None
    grade_level: Optional[str] = Field(None, max_length=20)
    email: Optional[str] = Field(None, max_length=100)
    phone: Optional[str] = Field(None, max_length=20)
    notes: Optional[str] = None
    active: Optional[bool] = None
    blocked_reason: Optional[str] = Field(None, max_length=200)

    @field_validator('role', mode='before')
    @classmethod
    def validate_role(cls, v):
        """Validate and convert role to enum."""
        if v is None:
            return None
        if isinstance(v, str):
            return BorrowerRole(v)
        return v


class BorrowerResponse(BorrowerBase, TimestampMixin):
    """Schema for borrower response."""

    id: int
    full_name: str
    barcode: str
    active: bool
    blocked_reason: Optional[str]

    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "example": {
                "id": 1,
                "borrower_id": "101",
                "first_name": "Amira",
                "last_name": "BENALI",
                "full_name": "Amira BENALI",
                "role": "student",
                "class_id": 1,
                "grade_level": "CP",
                "barcode": "%101",
                "active": True,
                "blocked_reason": None,
                "created_at": "2026-01-30T10:00:00",
                "updated_at": "2026-01-30T10:00:00",
            }
        }
    )


class BorrowerSummary(BaseModel):
    """Summary schema for borrower (for lists)."""

    id: int
    borrower_id: str
    full_name: str
    role: BorrowerRole
    class_id: Optional[int]
    active: bool
    blocked_reason: Optional[str]

    model_config = ConfigDict(from_attributes=True)


class BorrowerDetailed(BorrowerResponse):
    """Detailed borrower schema with statistics."""

    current_loans_count: int = Field(0, description="Number of items currently on loan")
    total_checkouts: int = Field(0, description="Total number of checkouts (all time)")
    overdue_count: int = Field(0, description="Number of overdue items")
    loan_limit: int = Field(2, description="Maximum number of items that can be borrowed")
    loan_limit_warning: int = Field(1, description="Soft warning threshold for loans")
    class_name: Optional[str] = Field(None, description="Class name (if student)")
    homeroom_teacher: Optional[str] = Field(None, description="Homeroom teacher for the class")

    model_config = ConfigDict(from_attributes=True)
